### PR TITLE
Better and more complete cleanup for CI jobs

### PR DESCRIPTION
**Pull request by potiuk**
Some of the recent refactors caused a problem that mypy jobs started to return errors likely resulting from a bad cache volume remaining after the jobs completed.

This PR extracts cleanup step to a separate, composite action and uses this action everywhere, including pruning docker including all containers and volumes to avoid the issue.

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: .github/actions/cleanup/action.yml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+#
+---
+name: 'Cleanup docker and repo'
+description: 'Cleans up docker and repository'
+runs:
+  using: "composite"
+  steps:
+    - name: "Cleanup docker"
+      shell: bash
+      run: docker system prune --all --force --volumes
```
Comment (potiuk) [inline .github/actions/cleanup/action.yml]: This `--volumes` here is probably the key to `mypy` failures.

**Review by aritra24 — APPROVED** [review]
Looks good to me

**Comment by potiuk** [comment]
Actually - we have to leave the old code - because it is needed to handle some "spilled" checkout problems and the shared action cannot be run before checkout - so we have chicken-egg problem :( .. This version is a more complete fix.

**Comment by potiuk** [comment]
The ARM CI failure might be ignored  - it's not relevant for the fix here.

**Review by uranusjr — APPROVED** [review]

**Review by dirrao — APPROVED** [review]

**Pull request merged**